### PR TITLE
Teach web-animations about matrixes.

### PR DESCRIPTION
**Pull request by shans**
Provide fallback to matrix interpolation when transform components are unavailable.

### CODE REVIEW

File: web-animations.js [hunk 1]
```diff
@@ -2092,52 +2092,380 @@ var transformREs = [
   buildMatcher('scaleY', 1, false, false, 1),
   buildMatcher('scaleZ', 1, false, false, 1),
   buildMatcher('scale3d', 3, false, false),
-  buildMatcher('perspective', 1, false, true)
+  buildMatcher('perspective', 1, false, true),
+  buildMatcher('matrix', 6, false, false)
 ];
 
+function decomposeMatrix(matrix) {
```
Comment (dstoc) [inline web-animations.js]: Let's wrap this so we don't have to allocate those utility function objects on every call, something like:

var decomposeMatrix = (function() {
  function utility {
  }
  function decomposeMatrix() {}
  return decomposeMatrix;
})();

File: web-animations.js [hunk 2]
```diff
@@ -2092,52 +2092,380 @@ var transformREs = [
...
+  // Therefore
+  //
+  // [A [0]]^-1 = [A^-1       [0]]
+  // [C  1 ]      [ -CA^-1     1 ]
+  function inverse(m) {
+    var iDet = 1 / determinant(m);
+    var a = m[0][0], b = m[0][1], c = m[0][2];
+    var d = m[1][0], e = m[1][1], f = m[1][2];
+    var g = m[2][0], h = m[2][1], k = m[2][2];
+    var Ainv = [[(e*k - f*h) * iDet, (c*h - b*k) * iDet, (b*f - c*e) * iDet, 0],
```
Comment (dstoc) [inline web-animations.js:2128]: use lowerCamel

File: web-animations.js [hunk 3]
```diff
@@ -2092,52 +2092,380 @@ var transformREs = [
...
+
+  if (m3d[0][3] != 0 || m3d[1][3] != 0 || m3d[2][3] != 0) {
+    rhs.push(m3d[0][3]);
+    rhs.push(m3d[1][3]);
+    rhs.push(m3d[2][3]);
+    rhs.push(m3d[3][3]);
+
+    var inversePerspectiveMatrix = inverse(perspectiveMatrix);
+    var transposedInversePerspectiveMatrix 
+        = transposeMatrix4(inversePerspectiveMatrix);
```
Comment (dstoc) [inline web-animations.js]: break after binary/assignment operators and

File: web-animations.js [hunk 4]
```diff
@@ -2092,52 +2092,380 @@ var transformREs = [
...
+  row[2] = combine(row[2], row[1], 1.0, -skew[2]);
+
+  scale.push(length(row[2]));
+  row[2] = normalize(row[2]);
+  skew[1] /= scale[2];
+  skew[2] /= scale[2];
+
+  var pdum3 = cross(row[1], row[2]);
+  if (dot(row[0], pdum3) < 0) {
+    for (i = 0; i < 3; i++) {
```
Comment (dstoc) [inline web-animations.js]: var i

File: web-animations.js [hunk 5]
```diff
@@ -2092,52 +2092,380 @@ var transformREs = [
...
+      matrix[i][j] *= scale[i];
+    }
+  }
+
+  return {t: 'matrix', d: [matrix[0][0], matrix[0][1],
+                           matrix[1][0], matrix[1][1],
+                           matrix[3][0], matrix[3][1]]};
+}
+
+function doItWithMatrices(from, to, f) {
```
Comment (dstoc) [inline web-animations.js]: do what?

File: web-animations.js [hunk 6]
```diff
@@ -2092,52 +2092,380 @@ var transformREs = [
...
+                toM.quaternion[i] * w);
+    }
+  }
+
+  var translate = interp(fromM.translate, toM.translate, f);
+  var scale = interp(fromM.scale, toM.scale, f);
+  var skew = interp(fromM.skew, toM.skew, f);
+  var perspective = interp(fromM.perspective, toM.perspective, f);
+
+  var result = composeMatrix(translate, scale, skew, quat, perspective);
```
Comment (dstoc) [inline web-animations.js]: just return composeMatrix(...)

File: web-animations.js [hunk 7]
```diff
@@ -2092,52 +2092,380 @@ var transformREs = [
...
+      out.push(doItWithMatrices(from.slice(i), to.slice(i), f));
+      return out;
     }
+
+    for (; i < from.length; i++)
+      out.push(interpTransformValue(from[i], {t: null, d: null}, f));
+
+    for (; i < to.length; i++)
+      out.push(interpTransformValue({t: null, d: null}, to[i], f));
+http://jsfiddle.net/pKdDe/
```
Comment (dstoc) [inline web-animations.js]: oops

File: web-animations.js (hunk 1 above)
Comment (shans) [inline web-animations.js]: done

File: web-animations.js (hunk 2 above)
Comment (shans) [inline web-animations.js:2128]: as discussed, this currently matches the submatrix A in the comments.

File: web-animations.js (hunk 3 above)
Comment (shans) [inline web-animations.js]: done and

File: web-animations.js (hunk 4 above)
Comment (shans) [inline web-animations.js]: done

File: web-animations.js (hunk 5 above)
Comment (shans) [inline web-animations.js]: it

File: web-animations.js (hunk 6 above)
Comment (shans) [inline web-animations.js]: done

File: web-animations.js (hunk 7 above)
Comment (shans) [inline web-animations.js]: indeed

**Pull request merged**